### ISSUE TITLE
Spec for server status protocol

### ISSUE THREAD

**Issue by Daenyth** [Cockatrice/Cockatrice#2474]
From #742, we need a spec for a server's information

Add a new command able to be sent to a server when not logged in
- Propose `Command_GetServerInfo`

A new response container for the useful information
- Propose `Response_ServerInfo_ServerStatus

Response should have fields useful to the UI:
- require reg to connect?
- reg is enabled?
- Current number of online users
- configured max online users

Any further suggestions?

**Comment by Daenyth** [comment]
#2201 becomes possible with this

**Comment by Daenyth** [comment]
- Response can also have a short length-limited (256?) text field for a description of the server
- Server should include required + supported features to connect

Should it include the version of the server?

**Comment by woogerboy21** [comment]
Any concerns with possible server attack vectors on something like this?

**Comment by skwerlman** [comment]
I can't think of a reason to include the server version, but I don't think there's any reason not to.

I think it should include whether the server has a scheduled shutdown pending, so people don't log on only to be told that the server is shutting down right now in the middle of their game. It'd be nice to see how long until the shutdown as well as the reason, but I don't think that part's as important.

I don't think there's any risk of attack here, but IANAE.

**Comment by ZeldaZach** [comment]
Some things I'd like to see this protocol have:

- server status (online, pending reboot, offline [aka no response])
- server requirements (clientId, mod log, registered user, etc)
- can you register on this server (independent of what the requirement field says)
- how many users are online this server
- how many users can this server support

**Comment by Daenyth** [comment]
> Any concerns with possible server attack vectors on something like this?

That was my reasoning for not necessarily including the version. Also the version isn't really helpful to users - can be a custom version or newer or older than client and none of it matters if feature set matches

**Comment by Daenyth** [comment]
Zach I think you just restated all the things I mentioned :P

**Comment by woogerboy21** [comment]
Identifying the scheduled shutdown time remaining would be new to me.  Is there a way to query the amount of time left on a timer?

**Comment by tooomm** [comment]
>#2201 becomes possible with this

Should be the same for #1702 then

>I think it should include whether the server has a scheduled shutdown pending, so people don't log on only to be told that the server is shutting down right now in the middle of their game. It'd be nice to see how long until the shutdown as well as the reason, but I don't think that part's as important.

Well, they get the `scheduled shutdown` message right after login right now I think?
So maybe not needed, still nice to have though...

**Comment by ZeldaZach** [comment]
Another thing we can respond with is a "support contact location" (required) and a "website" (optional) so we can return

```
{
   "server_name": WoogerWorks
   "server_addr": cockatrice.woogerworks.com
   "server_status": ONLINE
   "is_registration_open: true
   "server_required_features": [
        mod_lookup,
        clientID
    ]
   "server_ports": [
        4747,
        433
    ]
   "server_current_users": 460
   "server_max_users": 1200
   "server_contact": example@woogerworks.com
   "server_website": www.woogerworks.com
}
```

**Comment by Daenyth** [comment]
I don't think we can make contact required - consider the unconfigured servatrice build we want people to run for lan mode. Otherwise, a great addition

**Comment by tooomm** [comment]
Here are some related thoughts that might get lost in the closed issue:
>- try to ping each server name, and display the result;
>- (advanced, needs proper design): connect without logging in and ask the server for basic informations (requires registration / accepts new registrations / has free slots for new users / current num of users and games / planned shutdown / whatever )

About pinging there are some more infos in https://github.com/Cockatrice/Cockatrice/issues/2226#issuecomment-271102972 and following comments.

**Comment by tooomm** [comment]
It would be helpful if a client knows if a server allows for registration from client directly. Otherwise it should show a message and at best allows to display an url to a webpage for more information or registration there (like e.g. chickatrice does it).

Background:
@lachee seems to do registration on his webpagae only (for chickatrice) to prevent spam/automated registrations.
Zach had the idea to create an in-client capture to prevent that.
